### PR TITLE
Pandas based linear regression

### PR DESCRIPTION
**Pull request by henrydavidge**
Signed-off-by: Henry D <henrydavidge@gmail.com>

## What changes are proposed in this pull request?
(Details)

## How is this patch tested?
- [ ] Unit tests
- [ ] Integration tests
- [ ] Manual tests

(Details)

### CODE REVIEW

File: python/glow/gwas/linear_regression.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    So, if a matrix's indices are `sg` (like the X matrix), it has one row per sample and one column per genotype.
+    '''
+    X = np.column_stack(genotype_df['values'].array)
+
+    with oe.sharing.shared_intermediates():
+        XdotY = Y.T @ X - _einsum('cp,sc,sg,sp->pg', QtY, Q, X, Y_mask)
+        XdotY = Y.T @ X - _einsum('cp,sc,sg,sp->pg', QtY, Q, X, Y_mask)
+        QtX = _einsum('sc,sp,sg->pgc', Q, Y_mask, X)
+        XdotX_reciprocal = 1 / (_einsum('sp,sg,sg->pg', Y_mask, X, X) -
+                                _einsum('pgc,pgc->pg', QtX, QtX))
```
Comment (kianfar77) [inline python/glow/gwas/linear_regression.py:79]: Again I prefer we do not call this XdotX.

File: python/glow/gwas/linear_regression.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    c: covariate
+
+    So, if a matrix's indices are `sg` (like the X matrix), it has one row per sample and one column per genotype.
+    '''
+    X = np.column_stack(genotype_df['values'].array)
+
+    with oe.sharing.shared_intermediates():
+        XdotY = Y.T @ X - _einsum('cp,sc,sg,sp->pg', QtY, Q, X, Y_mask)
+        XdotY = Y.T @ X - _einsum('cp,sc,sg,sp->pg', QtY, Q, X, Y_mask)
+        QtX = _einsum('sc,sp,sg->pgc', Q, Y_mask, X)
```
Comment (kianfar77) [inline python/glow/gwas/linear_regression.py:77]: Can we make this ->pcg instead of pgc to match QtX?

**Review by kianfar77 — CHANGES_REQUESTED** [review]
Clean way of masking! I had some comments/questions.

File: python/glow/gwas/linear_regression.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    '''
+    Residualize a matrix in place using an orthonormal basis. The residualized matrix
+    is returned for easy chaining.
+    '''
+    M -= Q @ (Q.T @ M)
+    return M
+
+
+@typechecked
+def _linear_regression_inner(genotype_df: pd.DataFrame, Y: NDArray[(Any, Any), np.float64],
```
Comment (kianfar77) [inline python/glow/gwas/linear_regression.py:115]: Can you use a name other than `genotype_df` here as it is used for the spark dataframe before?

File: python/glow/gwas/linear_regression.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+                             YdotY: NDArray[(Any), np.float64], Y_mask: NDArray[(Any, Any),
+                                                                                np.float64],
+                             Q: NDArray[(Any, Any), np.float64], dof: int,
+                             phenotype_names: pd.Series) -> pd.DataFrame:
+    '''
+    Applies a linear regression model to a block of genotypes.
```
Comment (kianfar77) [inline python/glow/gwas/linear_regression.py:121]: Can you add a comment here and perhaps in the `linear_regression` function that the problem after projection covariates is performing multiple single variable regressions in parallel? The notation in linear algebra can be easily confused with where X is a vector for multi-variate regression.

File: python/environment.yml [hunk 5]
```diff
@@ -8,7 +8,9 @@ dependencies:
   - click=7.1.1 # Docs notebook source generation
   - databricks-cli=0.9.1 # Docs notebook source generation
   - jinja2
+  - jupyter
```
Comment (kianfar77) [inline python/environment.yml:11]: Is this related to this PR?

**Review by kianfar77 — COMMENTED** [review]
Great! Just a few minor comments.

File: python/environment.yml (hunk 5 above)
Comment (henrydavidge) [reply python/environment.yml:11]: No, it's very useful for testing though

File: python/glow/gwas/linear_regression.py (hunk 4 above)
Comment (henrydavidge) [reply python/glow/gwas/linear_regression.py:121]: Good suggestion

**Review by kianfar77 — APPROVED** [review]
Thanks
